ReadMany Pipeline

simorenoh opened the pull request:
# ReadMany Pipeline

This PR adds the changes for the client engine to work for ReadMany requests passing in a list of item identities (id, pk tuples) for querying in a more optimized manner. The main logic here is to group these item identities based on their relevant physical partitions in order to send partitioned queries targeting the items in those partitions, reducing overall time spent on the wire.

## Design and new classes

### `QueryPipeline::for_read_many`
In this case, since the pipeline is not taking in a singular query but rather a list of item identities, along with other partition key information from the container, it is necessary to come up with a different implementation of the usual QueryPipeline that can take in these new arguments. We opted for making a different instance of QueryPipeline that can be build using these specific settings. This pipeline also adds the implementations for the different methods we use in order to generate `QueryChunk`s.

### ReadMany producer strategy
Along the same lines, the other producers were used to dealing with queries that run the same across physical partitions - in this case, the query will be entirely different and physical partitions might even come up more than once (since we execute in batches of 1000 items per query, if we were doing a ReadMany with 2000 items all scoped to the same physical partition, we would have two sets of queries targeting the same physical partition) - so a new producer was necessary to deal with this, since the existing ones cycle through partition key range ids available and not just a subset/ a specific one targeted to a specific query.

### `QueryChunk` and `QueryChunkState`
Two new classes that help us with keeping track of the different queries being built and the state of whether they have executed yet or not.

## Things currently missing in this implementation

- [x] - We currently only handle `String` pk values - I will be porting over the PartitionKeyValue implementation from the Rust SDK in order to ensure we can handle the other relevant types for EPK hashing logic. - This was discussed in the PR comments and should be ok for now.
- [x] - Returning fully ordered results - for now I just wanted to get the core changes out so that we could start testing and seeing the different points within the client engine where pieces might need to move. Since the final ordering would be at the end, I wanted to make sure to leave this until after we could successfully run everything. - Addressed.
- [x] - Local tests - my main focus currently is validating the end-to-end flow against the Go SDK, local tests for the different pieces are secondary to that. - Currently in progress.

simorenoh committed:
read many changes

simorenoh committed:
read many changes

simorenoh committed:
ran `just headers`

simorenoh committed:
Update mod.rs

simorenoh committed:
remove comment

simorenoh committed:
test updates

simorenoh committed:
formatting rules from pipeline

simorenoh committed:
more fmt

simorenoh committed:
Update mod.rs

simorenoh committed:
slight changes

simorenoh committed:
missing changes from new responses

simorenoh committed:
small updates from testing

simorenoh committed:
refactoring logic from new class to previous querypipeline

simorenoh committed:
Update engine.go

simorenoh committed:
Update mod.rs

Pilchie commented on azure_data_cosmos_engine/src/query/producer/mod.rs:
```diff
@@ -51,6 +59,49 @@ pub fn create_partition_state(
+
+fn create_query_chunk_query(
+    key: &str,
+    map: &HashMap<String, Vec<(usize, String, String)>>,
```
Wondering why take the `key` and the `map` here instead of just taking a borrow of the `identities` `Vec` directly?

Pilchie commented on azure_data_cosmos_engine/src/query/producer/mod.rs:
```diff
@@ -100,15 +151,23 @@ impl ItemProducer {
+
     /// Gets the [`DataRequest`]s that must be performed in order to add additional data to the partition buffers.
-    pub fn data_requests(&mut self) -> Vec<DataRequest> {
+    pub fn data_requests(&mut self) -> crate::Result<Vec<DataRequest>> {
```
Why make this change?

simorenoh replied:
was cherry-picking changes from Ashley's hybrid search PR - it is to accommodate for the Hybrid state being introduced since it doesn't follow the same logic of using `Ok()` on the other strategies' requests() method, and as a result the outer method has to return the same type - @analogrelay do correct me if I'm wrong

analogrelay replied:
Yep. @simorenoh is exactly right, the Hybrid strategy can fail here (though only due to the SDK calling things in the wrong order)

Pilchie commented on azure_data_cosmos_engine/src/query/producer/state.rs:
```diff
@@ -64,10 +64,72 @@ impl PartitionState {
+#[derive(Debug)]
+pub struct QueryChunkState {
+    /// The index of the partition in the query chunks list used by the pipeline.
+    pub index: usize,
```
Where is this used?

analogrelay replied:
Might be copied from `PartitionState`, which doesn't really need it either tbh :).

simorenoh replied:
@analogrelay I actually ended up using this as a sort of `id` identifier for any given state since in ReadMany a partition key range id can be repeated more than once (and as such can't reliably be used to check a state)

analogrelay replied:
Yep, this totally makes sense now that I get that you can have multiple chunks for a single PK Range.

Pilchie commented on azure_data_cosmos_engine/src/query/mod.rs:
```diff
@@ -87,6 +87,31 @@ impl PartitionKeyRange {
+    derive(pyo3::FromPyObject),
+    pyo3(from_item_all)
+)]
+#[serde(rename_all = "camelCase")]
```
```suggestion
#[serde(rename_all = "PascalCase")]
```

And then you can maybe get rid of the renames below?

analogrelay replied:
The `Deserialize` implementation (which I assume is copied from one of our other models) is for Python compatibility (i.e. to "deserialize" an `ItemIdentity` from a Python object with `PartitionKeyValue` and `ID` properties).. We aren't prioritizing Python yet so I'd recommend we just get rid of it. We can always add it back in later, when we know more about how we'll structure the APIs. So you can just remove:

* The `Deserialize` from the `#[derive]` attribute
* The entire `#[cfg_attr]` attribute
* All `#[serde]` attributes from the type and fields.

analogrelay replied:
Oop! I see I'm wrong about that, it's used at the FFI level (Go passes a JSON-serialized value in here). If you're going to do that, I'd suggest keeping everything `camelCase` though to be consistent with other Cosmos DB backend APIs.

Pilchie commented on azure_data_cosmos_engine/src/query/producer/mod.rs:
```diff
@@ -51,6 +59,49 @@ pub fn create_partition_state(
+            .keys()
+            .next()
+            .cloned()
+            .expect("Expected exactly one key in each hashmap");
```
A `HashMap` that only ever has one key seems like a weird choice. Why not just a tuple instead?

Pilchie commented on azure_data_cosmos_engine/src/hash.rs:
```diff
@@ -179,8 +191,8 @@ impl PartitionKeyValue {
-    kind: PartitionKeyKind,
-    version: u8,
+    kind: &PartitionKeyKind,
+    version: u32,
```
Why make these changes?

analogrelay replied:
Yeah, we could probably leave this as `u8`, unless we suspect it's likely we'll have 253 more Partition Key Versions appear _before_ we're willing to take a breaking change in a major version :).

I do think we should use a exact-size integer (`u8`/`u32` as opposed to `usize`) to keep the types clear on the FFI boundary.

analogrelay replied:
Also, there's no reason to use `&PartitionKeyKind` here.  The `PartitionKeyKind` enum is small enough that it's no bigger than a pointer. Instead, take `PartitionKeyKind` AND add `Clone, Copy` to the `#[derive]` attribute for `PartitionKeyKind`. This tells Rust that `PartitionKeyKind` is just a simple piece of data that can be copied "for free" (i.e. copying is no more costly than passing a pointer to it).

In general, any enum that is just a simple wrapper around an integer should be `Clone, Copy`, and then you basically never need to pass references, you can just pass it by value.

simorenoh replied:
did all of these changes, reverted the u32 - @Pilchie reason for u32 was Go was directly passing int32 so I just wanted to align it, but it makes more sense with what Ashley's saying as well to just change it on the Go side.

Pilchie commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,98 @@
+    pub fn requests(&mut self) -> Vec<DataRequest> {
+        // Here is where we would create DataRequests for the current query chunk.
+
+        // In the unordered strategy, we simply return the first partition key range's request.
```
This isn't the unordered strategy anymore.

Pilchie commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,98 @@
+        tracing::debug!(query_chunk_index = ?self.current_query_chunk_index, "getting requests for current query chunk");
+        tracing::debug!(query_chunk_states = ?self.query_chunk_states.get(self.current_query_chunk_index), "current query chunk state");
+        while requests.is_empty() {
+            // If there are no more partitions, return None.
```
This doesn't return `None`.

analogrelay replied:
Yeah, that _was_ the contract of `requests` before, but I refactored it recently. When there are no requests, just return an empty `Vec`.

Pilchie reviewed (commented):
No tests?

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,98 @@
+use super::{create_query_chunk_states, state::QueryChunkState};
+
+pub struct ReadManyStrategy {
+    pub query_chunks: Vec<HashMap<String, Vec<(usize, String, String)>>>,
```
We should avoid using anonymous tuples unless they're very straightforward. Adding a simple struct with public named fields is an order of magnitude more readable with only small amount of extra code. This could be a `Vec<QueryChunk>`, and `QueryChunk` could be:

```rust
pub(crate) struct QueryChunk {
    pub partition_key: String,
    pub items: Vec<QueryChunkItem>
}

pub struct QueryChunkItem {
    index: usize,
    pkrange_id: String,
    query: String
}
```

It's important to note that types are very cheap in Rust. There's no risk of increasing Heap allocations, and performance costs for constructing structs are identical to tuples. Tuples are reasonable for one-off private function args/returns, but I try to avoid them for more complicated things.

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,98 @@
+pub struct ReadManyStrategy {
+    pub query_chunks: Vec<HashMap<String, Vec<(usize, String, String)>>>,
+    pub current_query_chunk_index: usize,
+    pub query_chunk_states: Vec<QueryChunkState>,
```
Could this be merged into the `query_chunks` property and just store a single list of `QueryChunkState` which includes both the relevant info AND the pagination state?

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,98 @@
+            current_query_chunk_index: 0,
+            query_chunk_states: query_chunk_states,
+            items: VecDeque::new(),
+            result_shape: QueryResultShape::RawPayload,
```
No real need to store this. Some other strategies still do that but it's leftover from prior refactorings. The only real reason to "store" a `QueryResultShape` is if a single producer could have different shapes. For example, the Unordered strategy could have `RawPayload` or `ValueAggregate`.

Instead, where you call `self.result_shape.results_from_slice`, you can just call `QueryResultShape::RawPayload.results_from_shape` directly.

simorenoh replied:
done

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,98 @@
+        // Once that partition is exhausted, we remove it from the list and return the next one.
+        let mut requests = Vec::new();
+        tracing::debug!(query_chunk_index = ?self.current_query_chunk_index, "getting requests for current query chunk");
+        tracing::debug!(query_chunk_states = ?self.query_chunk_states.get(self.current_query_chunk_index), "current query chunk state");
```
nit: I'd merge these trace events. Plus, I believe the state contains the index already

simorenoh replied:
yeah most of the traces are just for my personal debugging - haven't replied to most comments since I'm still validating the changes/ making changes needed to work

analogrelay replied:
No problem! I figured that was the case.

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,98 @@
+        let mut requests = Vec::new();
+        tracing::debug!(query_chunk_index = ?self.current_query_chunk_index, "getting requests for current query chunk");
+        tracing::debug!(query_chunk_states = ?self.query_chunk_states.get(self.current_query_chunk_index), "current query chunk state");
+        while requests.is_empty() {
```
By my reading, this will only ever produce 0 or 1 requests. Either it will return a single query for the active chunk, or no query at all. That's a reasonable start for a v1 MVP, but we could absolutely return several queries here, one for each chunk. That would allow the SDK to parallelize the requests, if it chooses to.

We don't do that in the Unordered strategy because the user is paginating results and may _stop_ requesting new data. We don't want to have pre-fetched a bunch of data the user never ends up iterating to. That's not the case here though, we're buffering all the results and returning one big list. So it's safe to tell the SDK it can issue several simultaneous requests (which is what we do in the Streaming strategy, for example).

analogrelay commented on azure_data_cosmos_engine/src/query/producer/state.rs:
```diff
@@ -64,10 +64,72 @@ impl PartitionState {
+}
+
+#[derive(Debug)]
+pub struct QueryChunkState {
```
This is largely the same as `PartitionState`, but with an additional field for the query. How about we merge it with `PartitionState` and just make the `query` an `Option<String>`? It would add a small amount of extra memory to the `PartitionState` type when no query is needed (to store the `None`), but I'm really not worried about that yet (and if we absolutely need to worry about it, there are ways to avoid that).

simorenoh replied:
the only other thing here that is different/ that gets used differently is the index - since we can't use the partition key range id to map to a state (more than one query chunk state can have the same partition key range id) I chose to use the index (which gets assigned as we create the query chunk states sequentially) - and that is what we use in `provide_data` to check the individual state and update its continuation. not sure if that's enough to warrant the separate class since the states will be handled a bit differently or if there's some way we could accommodate for this across all partition states currently being used.

analogrelay replied:
That makes sense. It's not unlike what I ended up doing for Hybrid. Hybrid queries also manage multiple separate queries per PK Range (the component queries), and I handled that by splitting the `u64` request ID into a `u32` for the component query index (probably overkill and I could use a `u8` or `u16`, but it should be OK for now), and a `u32` for the page number.

I'm OK with keeping this separate state type around in that case. If this  is only used by ReadMany, maybe move it to `read_many.rs`? If that file is feeling big, you can always change it to a directory-based module (`read_many/mod.rs`) with further submodules. But don't be afraid of a single `.rs` file containing multiple types and getting fairly long. That's quite common in Rust.

analogrelay commented on azure_data_cosmos_engine/src/query/pipeline.rs:
```diff
@@ -295,14 +297,129 @@ impl QueryPipeline {
+    pub fn for_read_many(
+        item_identities: impl IntoIterator<Item = ItemIdentity>,
+        pkranges: impl IntoIterator<Item = PartitionKeyRange>,
+        pk_kind: &str,
```
nit: `pk_kind` should be `PartitionKeyKind` here. Do the string parsing in the C binding rather than here. This API is intended to be a more standard Rust API, and in that situation you'd expect the caller to provide the enum rather than it's string implementation. The fact that we pass this as a string is a concern of the C bindings.

analogrelay commented on azure_data_cosmos_engine/src/query/pipeline.rs:
```diff
@@ -295,14 +297,129 @@ impl QueryPipeline {
+
+    // Groups items by their partition key range ID efficiently while preserving original order.
+    // Returns a mapping of partition key range IDs to lists of tuples containing the original index, item ID, and partition key value.
+    fn partition_items_by_range(
```
nit: In Rust you should generally only put a function in an `impl` block if:

* It accepts `Self` (or `&Self`, `&mut Self`, etc) as a parameter, and/or
* It returns `Self` (or `Result<Self>`, `Option<Self>`, etc.)

This should just be a bare `fn` outside the `impl` block. It's quite idiomatic to have `fn`s directly inside a module, rather than grouping them into types (like static classes in .NET/Java/etc.)

analogrelay replied:
As you refactor this based on my suggestion above, keep that rule in mind. If it doesn't use or return `Self`, just make it a top-level `fn` in the module.

analogrelay commented on azure_data_cosmos_engine/src/query/pipeline.rs:
```diff
@@ -295,14 +297,129 @@ impl QueryPipeline {
+        // Create query chunks from the partitioned items, splitting total list into 1000 max item queries.
+        // Each chunk is represented as vector of mappings of partition key range IDs to lists of tuples containing the original index, item ID, and partition key value.
+        let query_chunks = Self::create_query_chunks_from_partitioned_items(&items_by_range);
+        tracing::debug!(?query_chunks, "created query chunks from partitioned items");
```
I'd take both of these functions (`partition_items_by_range` and `create_query_chunks_from_partitioned_items`) and make them into a single constructor on a new `QueryChunk` type (as mentioned above), with this signature:

```rust
impl QueryChunk {
    pub fn from_identities(item_identities: Vec<ItemIdentity>, pkranges: Vec<PartitionKeyRange>, pk_kind: PartitionKeyKind, pk_version: u32) -> Vec<Self> // Or Result<Vec<Self>> if necessary
}
```

Within `from_identities`, you're welcome to call out to other `fn`s (which may not need to live inside the `impl` itself. But I think it makes sense to group all the chunking logic into a type that represents a single chunk.

simorenoh replied:
I moved everything over to `mod.rs` to accompany the QueryChunk definition and implementation - but I feel like it makes sense from a developer perspective and a maintainability perspective for us to keep the split methods on everything they do, otherwise it just becomes one big piece of code with no separation other than comments... if you feel like this is still wrong, definitely let me know and I'll reformat it into the `from_identities` method

analogrelay commented on azure_data_cosmos_engine/src/query/pipeline.rs:
```diff
@@ -295,14 +297,129 @@ impl QueryPipeline {
+        pk_kind: &str,
+        pk_version: u32,
+    ) -> HashMap<String, Vec<(usize, String, String)>> {
+        // TODO: Partition key values here are all currently strings - we need the same sort of PartitionKeyValue
```
Yep, this is on my mind too. I think we should probably just represent a Partition Key as a JSON array at the FFI boundary. That way we have a clear format that supports encoding all the relevant type information. So, if you want to pass a string PK, you'd pass `["partition0"]`, or `[42]`, or `[null]`, etc. Then we can deserialize that. It would mean the Go code has to "serialize" the PK to JSON and we have to "deserialize" it, but that's probably better than trying to work out some kind of data structure to share between Go and Rust.

Anyway, it's fine to expect strings for now.

analogrelay commented on azure_data_cosmos_engine/src/query/pipeline.rs:
```diff
@@ -295,14 +297,129 @@ impl QueryPipeline {
+
+            let epk_range_string =
+                get_hashed_partition_key_string(&[pk_value_val], &pk_kind_enum, pk_version);
+            let epk_range = QueryRange::new(&epk_range_string, &epk_range_string, true, true);
```
This results in one unnecessary clone. The `get_hashed_partition_key_string` function is already returning a `String`. But then `QueryRange::new` takes `&str`, and uses `.into()` (which will "clone" the string into a new heap-allocated `String`). So, you end up creating 3 `String`s on the heap: The one `get_hashed_partition_key_string` creates, and two clones in `QueryRange::new`. Then you destroy the original `String`. Instead you can just make one clone and then **move** the two clones into the `QueryResult`.

A constructor function (`QueryRange::new`) isn't really necessary here. For internal types like this, it's also just fine (and completely idiomatic) to construct the struct directly. The fields are public, so that works just fine and you can replace what you have with this (and remove the `QueryRange::new` function):


```suggestion
            let epk_range = QueryRange {
                min: epk_range_string.clone(), // Clones epk_range_string and moves the clone in
                max: epk_range_string, // Move the original string in. 
                min_inclusive: true,
                max_inclusive: true,
            }
```

We could probably also get rid of the `.clone` needed so that min and max can share the same value, but I'm not as worried about optimizing that right now.

analogrelay commented on cosmoscx/src/pipeline.rs:
```diff
@@ -79,6 +83,58 @@ pub extern "C" fn cosmoscx_v0_query_pipeline_create<'a>(
+/// - `pk_version`: The partition key version.
+#[no_mangle]
+pub extern "C" fn cosmoscx_v0_readmany_pipeline_create<'a>(
+    item_identities: Str<'a>,
```
Hmm... Not sure about this, it adds a fair bit of extra complexity to the FFI layer since the caller needs to take their list of (string, string) pairs and serialize them into a single string. We could instead take _two_ `Slice<'a, Str<'a>>` parameters (`ids`, and `partition_keys`) which are expected to line up exactly (The identity for item with index `x` is `(ids[x], partition_keys[x])`). That's a clumsy API to expose to end-users, but I'm not at all worried about using it at this internal FFI boundary. We'll control the caller and can make sure it does the right thing.

Having said that, we're in early days for this feature, and this API is internal and can be changed later without customer impact. I'm comfortable keeping this for now to save time and refactoring later.

analogrelay replied:
So, to be clear, let's leave this and revisit using two lists later. The Go wrapper carries it's own version of the engine binary with it, so it's safe for us to change this later as long as we make the change in both the engine and update the wrapper at the same time.

analogrelay commented on cosmoscx/src/pipeline.rs:
```diff
@@ -148,8 +220,11 @@ pub extern "C" fn cosmoscx_v0_query_pipeline_run(
     ) -> Result<Box<PipelineResult>, azure_data_cosmos_engine::Error> {
+        tracing::debug!("cosmoscx_query_pipeline_run starting");
         let pipeline = unsafe { Pipeline::unwrap_ptr(pipeline) }?;
+        tracing::debug!("unwrapped pipeline");
```
Probably unnecessary to have in the final code. Or should at least be demoted to `tracing::trace!`

analogrelay reviewed (changes requested):
Definitely some things to touch up, but this is looking pretty good! I see some stuff that I also have in my Hybrid PR (I assume you've done some rebasing/cherry-picking). Let's try to get that PR in so you can rebase and have a clean PR with just your changes in it.

simorenoh committed:
Merge branch 'main' into readmany-client-engine

simorenoh committed:
working implementation with partitioned container

missing final ordering on last pipeline.run() call, popping items based on their index within the identities list

simorenoh committed:
addressing some comments

simorenoh committed:
oops

simorenoh committed:
working sorting logic

simorenoh committed:
remove extra logs, more comments

simorenoh committed:
test updates

simorenoh committed:
add readmany strategy test

Pilchie commented on azure_data_cosmos_engine/src/query/mod.rs:
```diff
@@ -87,6 +89,144 @@ impl PartitionKeyRange {
+}
+
+impl QueryChunk {
+    pub fn from_identities(item_identities: Vec<ItemIdentity>, pkranges: &mut Vec<PartitionKeyRange>, pk_kind: PartitionKeyKind, pk_version: u8) -> Vec<QueryChunk> {
```
```suggestion
    pub fn from_identities(item_identities: Vec<ItemIdentity>, pkranges: &mut Vec<PartitionKeyRange>, pk_kind: PartitionKeyKind, pk_version: u8) -> Vec<Self> {
```

Nit: This makes the diff smaller if you rename the type.

Pilchie commented on azure_data_cosmos_engine/src/query/mod.rs:
```diff
@@ -87,6 +89,144 @@ impl PartitionKeyRange {
+    items_by_partition: &HashMap<String, Vec<(usize, String, String)>>,
+) -> Vec<QueryChunk> {
+    let mut query_chunks: Vec<QueryChunk> = Vec::new();
+    let max_items_per_query = 1000;
```
Shouldn't we also be limiting the chunk size if the query string gets to be too long for the service to accept?

analogrelay replied:
We probably should, yeah. Maybe, instead of stepping by `max_items_per_query`, you iterate item by item building a chunk and then cut off that chunk and move to a new one whenever the chunk reaches a certain query text size or item count.

analogrelay replied:
That would also mean building up the query as we chunk, but that should be fine.

simorenoh replied:
I ran some initial tests with Python and I'm not sure this is something we should be worried about with batches of 1000 items - I was able to get a query text of length 310k running first try and returning results using the format we want ("SELECT * FROM c WHERE ( (c.id = id AND c.pk=pk) OR (c.id=.... ) )". Subtracting the characters that are always included in the definition and the 999 OR statements with spaces around them, you still have around 305k characters to use for each term, which works out to 305 characters per inner term. That means each inner term could be something like

(c.id = 'Microsoft Corporation is an American multinational technology conglomerate headquartered in Redmond, Washington.' AND c.longestPartitionKeyPathInTheHistoryOfTheWorld = 'Founded in 1975, the company became influential in the rise of personal computers through software like Windows.') - and you would still have leftover characters

You would need some very long id values/ pk path/ pk value to even get close. And 310k is probably not even the limit on the service side either.

Pilchie replied:
Fair - I believe the limit is something like 1MB.  @FabianMeiswinkel can probably confirm.

analogrelay replied:
> c.longestPartitionKeyPathInTheHistoryOfTheWorld 

Challenge accepted. 😈 

---

Yep, I think this is reasonable. If we believe 1000 items is a good batch size (and I think it does make sense), then we should be safe on the query size.

@simorenoh can you emit a `debug!` trace after generating the query that includes the query length? (Include it as a "property", something like `tracing::debug!(query_len = query.len(), "generated query");`). That way we have a way to diagnose potential issues there.

Pilchie commented on azure_data_cosmos_engine/src/query/pipeline.rs:
```diff
@@ -332,7 +334,7 @@ impl QueryPipeline {
             // Termination MUST come from the pipeline, to ensure aggregates (which can only be emitted after all data is processed) work correctly.
             if result.terminated {
-                tracing::trace!("pipeline node terminated the pipeline");
+                tracing::debug!("pipeline node terminated the pipeline");
```
Do we want to keep this as `debug` or will it be too noisy?

simorenoh replied:
oh good catch - need to do another clean of tracings

analogrelay replied:
Yeah, let's shift it back to `trace`

simorenoh committed:
fmt

Pilchie commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,110 @@
+        let query_chunk_state = self
+            .query_chunk_states
+            .iter_mut()
+            .find(|state| state.index == request_id as usize)
```
How big do we expect this to be? Should we consider a map to make finding this easier?

analogrelay replied:
If the index is sequential and used as the request ID, couldn't we just index into the list directly here, rather than using a find?

simorenoh replied:
applied this logic

Pilchie commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,110 @@
+        let _ = query_chunk_state;
+
+        // if we're all done, we can sort the final list of items
+        if self.query_chunk_states.iter().all(|state| state.done()) {
```
Similarly, if we expect there to be a significant number of elements in `query_chunk_states` consider instead having a count of remaining items that you can decrement as each finishes and then just check for 0 remaining here.

analogrelay replied:
Good idea, I do that in the Hybrid producer as well

simorenoh replied:
applied this logic

Pilchie reviewed (commented):
LGTM, though I'll defer to @analogrelay for final signoff.

analogrelay commented on azure_data_cosmos_engine/src/query/producer/mod.rs:
```diff
@@ -56,6 +62,37 @@ pub fn create_partition_state(
+
+    let conditions = query_chunk_items
+        .iter()
+        .map(|item| format!("( c.id = '{}' )", item.id))
```
Could this be a `SELECT * FROM c WHERE c.id IN ('a', 'b', 'c', 'd', ...)` query? That would, I think, shave some characters off the final query text length by avoiding repetition of `c.id = ` and `OR`s

simorenoh replied:
yes this was initially the idea - but now it will be updated to carry (c.id = item.id AND c.pk = item.pk) OR (c.id = item.id AND c.pk = item.pk) OR... and so on, so this format will still make sense

analogrelay replied:
Ah right. Ok, that make sense to me.

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,123 @@
+                    request_id
+                ))
+            })?;
+        // Update the state and capture the done status before dropping the mutable borrow
```
nit: The more common pattern for a temporary mutable borrow followed by an immutable borrow is to use `{}` to create a seperate scope. For example:

```rust
{
    let mut_ref = &mut obj;
    // work with mutable reference
}

// Mutable borrow ended at the '}', so we can borrow immutably again
let r = &obj;
```

Though you may also find that sometimes it's not even necessary to do that. Rust recently added "non-lexical lifetimes" in which the compiler is smarter about detecting when you stop doing mutable things and releasing the mutable borrow then.

analogrelay replied:
(What you have is perfectly functional though, so this is non-blocking)

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,123 @@
+
+            let mut items_with_indices: Vec<(usize, QueryResult)> = self
+                .items
+                .drain(..)
```
I would use `into_iter()` instead of `drain(..)` here. The behavior of the two is slightly different.

1. `drain` takes a `&mut` reference to the vector, and pops all the items out of the vector, basically the same as calling `.pop` in a loop.
2. `into_iter` takes the vector by value (taking ownership of it), which means it doesn't have to keep the vector structure "intact" and has much more flexibility with how it cleans up the vector after you're done with it. Effectively, it can just disassemble the Vec into a start and end pointer and iterate through it.

The thing is, you can't just swap this call to `into_iter`, because `into_iter` requires that you own the vector but you're borrowing `self` and thus you can't just move `self.items`. But, there's a solve for that. You can call `let items = std::mem::take(self.items)` to move the entire vector out of `self.items` and into a local variable. That works because the `std::mem::take` moves the value out of a `&mut` reference **and** replaces it with the default value for that type (which has to implement `Default`). So you basically do a swap of `self.items` with an empty vec (which is very cheap to create and requires no heap allocations) and then you own the vector that actually has the items.

simorenoh replied:
wow this is a lot of Rust info, thanks Ashley!

analogrelay replied:
No problem! One of the tricky things to get a feel for in Rust is that _most_ of the time you assign values (`foo = bar`) you're moving a value _out_ of `bar`. That means either:

1. The compiler has to be able to prove you'll never try to access `bar` again (because its value is undefined)
2. OR, you have to put a valid new value in `bar` at the same time as you take the current value out. That way Rust knows that you can still access `bar`.

A number of functions in `std::mem` help you achieve #2, basically by using safely-written unsafe code to temporarily break rule #1 while it puts a new value in place. `replace` lets you provide a new value, `take` just uses the default value for the type (if there is one), etc.

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
```diff
@@ -0,0 +1,123 @@
+    let QueryResult::RawPayload(payload) = query_result else {
+        return None;
+    };
+    let json: serde_json::Value = serde_json::from_str(payload.get()).ok()?;
```
This is a small variation, but worth it. Instead of deserializing the whole payload into a `serde_json::Value` (which has to allocate new memory for the entire user's payload object), define a struct inside this function (that's totally valid in Rust btw, you can define a type inside a function, which is only usable within the function):

```rust
#[derive(Deserialize)]
struct ItemWithId {
    pub id: String
}
```

Then deserialize into that and read the `id` from that.

By using a type that only defines what we need, `serde_json` can stop parsing after it's read enough data to fill the type, and it doesn't have to allocate any memory for the other data in the JSON object.

simorenoh committed:
address comments

analogrelay commented on azure_data_cosmos_engine/src/query/pipeline.rs:
```diff
@@ -356,6 +359,39 @@ impl QueryPipeline {
+            QueryChunk::from_identities(item_identities, &mut pkranges, pk_kind, pk_version);
+        // Create the item producer for read many.
+        let producer = ItemProducer::read_many(query_chunks);
+        let pipeline: Vec<Box<dyn PipelineNode>> = Vec::new();
```
```suggestion
        let pipeline = Vec::new();
```

You shouldn't need the type annotation here, because `pipeline`'s type is fixed when you use it to construct the struct below.

analogrelay reviewed (commented):
Looking good! A few comments, and we should get an end-to-end test in place. I don't see any critical issues though 👍🏻

simorenoh committed:
include pk paths, run queries with pk values

simorenoh committed:
next set of comments

simorenoh committed:
fmt and tests

analogrelay commented on azure_data_cosmos_engine/src/query/producer/mod.rs:
```diff
@@ -64,33 +64,50 @@ pub fn create_partition_state(
+
+        format!("SELECT * FROM c WHERE ( {conditions} )")
+    } else {
+        // here we could have logic for HPK later down the line - for now we just do queries with only ID values
```
We should return an error here if we we're sure we can return an accurate value. Better to tell the user "HPK is not supported" instead of returning potentially incorrect data.

Of course, because this is Rust, that means you'll have to make this return a `Result` and update callers, but I think it's worth it.

simorenoh committed:
Update pipeline_read_many.rs

analogrelay commented on azure_data_cosmos_engine/src/query/producer/read_many.rs:
As discussed offline, let's remove this sort and make sure the API docs are clear that results are **not** guaranteed to be returned in the same order as the input tuples.

simorenoh committed:
remove all ordering logic and update tests

simorenoh committed:
add tracing for query length

simorenoh committed:
Update CHANGELOG.md

simorenoh committed:
add new error for HPK

simorenoh committed:
fmt

simorenoh committed:
Update mod.rs

simorenoh committed:
Merge branch 'main' into readmany-client-engine

simorenoh committed:
update changelog and Cargo.toml for release

simorenoh commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
+## 0.4.0 (2025-11-24)
+
+### Features Added
+* Added ability to execute `readMany` requests against the client engine. - [PR 53](https://github.com/Azure/azure-cosmos-client-engine/pull/53)
```
```suggestion
* Added ability to execute `readMany` requests against the client engine to read multiple items. Results are not guaranteed to be ordered. - [PR 53](https://github.com/Azure/azure-cosmos-client-engine/pull/53)
```

simorenoh committed:
Update CHANGELOG.md

simorenoh committed:
Update Cargo.lock

simorenoh committed:
go mod tidy

simorenoh committed:
updated tests to accommodate non-deterministic behavior

analogrelay commented on azure_data_cosmos_engine/tests/pipeline_read_many.rs:
```diff
@@ -98,38 +98,74 @@ pub fn read_many() -> Result<(), Box<dyn std::error::Error>> {
-            },
-        ],
+    
+    // Note: The order of partition results is non-deterministic and depends on which partition responds first
+    // We need to check for either ordering: even-first or odd-first
```
Small nit (not necessarily worth respinning): You could also sort the items by partition key, but I suppose that would be tricky given that we don't have objects, just `serde_json::Value`s.

simorenoh committed:
Update pipeline_read_many.rs

simorenoh committed:
update type
